lib/Map/TableDataSource.js

```diff
@@ -230,10 +230,81 @@ TableDataSource.prototype.loadUrl = function (url) {
 TableDataSource.prototype.loadText = function (text) {
     this.dataset.loadText(text);
     if (this.dataset && this.dataset.hasTimeData() && !defined(this.displayDuration)) {
+        // default duration calculation is for a data point to be shown for 1% of the total time span of the dataset
         var percentDisplay = 1.0;
         this.displayDuration = JulianDate.secondsDifference(this.dataset.getTimeMaxValue(), this.dataset.getTimeMinValue()) * percentDisplay / (60.0 * 100.0);
+        this._autoDuration = true; // Since no duration was provided, we will use the full interval between each data point and the next
     }
-    this.setDataVariable(this.dataVariable);
+
+    if (defined(this.dataVariable)) {
+        // user-specified variable
+        this.setDataVariable(this.dataVariable);
+    } else {
+        // otherwise pick the first one. Doing it here may simplify life.
+        this.setDataVariable(this.dataset.getDataVariableList(true)[0]);
+    }
+};
+
+/**
+ * Returns a cached, sorted, unique array of JulianDates corresponding to the time values of different data points. 
+ */
+
+TableDataSource.prototype.timeSlices = function() {
+    if (!this.dataset || !this.dataset.hasTimeData()) {
+        return undefined;
+    }
+    if (this._timeSlices) {
+        return this._timeSlices;
+    }
+    var pointList = this.dataset.getPointList();
+    var times = {};
+    for (var i = 0; i < pointList.length; i++) {
+        times[pointList[i].time.toString()] = true;
+    }
+    this._timeSlices = Object.keys(times).map(function(x) { return JulianDate.fromIso8601(x); } ); 
+
+    this._timeSlices = this._timeSlices.sort(function(a,b) {
+        if (a.equals(b))
+            return 0; // probably can't happen
+        return JulianDate.lessThan(a,b) ? -1 : 1;
+    });
+    return this._timeSlices;
+};
+
+/**
+ * Returns a { start, finish } pair of JulianDates corresponding to some pair of rows around this time. 
+ */
+TableDataSource.prototype.getTimeSlice = function(time) {
```

Comment:
Have you seen Cesium's `TimeInterval` and `TimeIntervalCollection` types?  I think they may simplify this logic quite a bit.